runtime/common/src/custom_xcm_barrier.rs

```diff
@@ -0,0 +1,121 @@
+use core::marker::PhantomData;
+use frame_support::{ensure, log, traits::Contains};
+
+use xcm::latest::{prelude::*, Instruction, Weight as XCMWeight};
+
+use xcm_executor::traits::ShouldExecute;
+
+use scale_info::prelude::boxed::Box;
+use sp_std::vec::Vec;
+
+pub trait ReserveAssetDepositedMatcher: Sync {
+	fn matches(&self, multi_asset: &MultiAsset) -> bool;
+}
+
+pub trait DepositAssetMatcher {
+	fn matches<'a>(
+		&self,
+		assets: &'a MultiAssetFilter,
+		beneficiary: &'a MultiLocation,
+	) -> Option<(u8, &'a [u8])>;
+}
+pub struct MatcherPair {
+	reserve_deposited_asset_matcher: Box<dyn ReserveAssetDepositedMatcher>,
+	deposit_asset_matcher: Box<dyn DepositAssetMatcher>,
+}
+
+impl MatcherPair {
+	pub fn new(
+		reserve_deposited_asset_matcher: Box<dyn ReserveAssetDepositedMatcher>,
+		deposit_asset_matcher: Box<dyn DepositAssetMatcher>,
+	) -> Self {
+		MatcherPair { reserve_deposited_asset_matcher, deposit_asset_matcher }
+	}
+
+	fn matches_reserve_deposited(&self, multi_asset: &MultiAsset) -> bool {
+		self.reserve_deposited_asset_matcher.matches(multi_asset)
+	}
+
+	fn matches_deposit_asset<'a>(
+		&'a self,
+		assets: &'a MultiAssetFilter,
+		beneficiary: &'a MultiLocation,
+	) -> Option<(u8, &'a [u8])> {
+		self.deposit_asset_matcher.matches(assets, beneficiary)
+	}
+}
+
+pub trait MatcherConfig {
+	fn get_matcher_pairs() -> Vec<MatcherPair>;
+	fn callback(length: u8, data: &[u8]) -> Result<(), ()>;
+	fn get_incoming_parachain_id() -> u32;
+}
+
+pub struct AllowUnpaidExecutionFromCustom<T, V> {
+	_phantom: PhantomData<(T, V)>,
+}
+impl<T: Contains<MultiLocation>, V: MatcherConfig> ShouldExecute
+	for AllowUnpaidExecutionFromCustom<T, V>
+{
+	fn should_execute<RuntimeCall>(
+		origin: &MultiLocation,
+		instructions: &mut [Instruction<RuntimeCall>],
+		_max_weight: XCMWeight,
+		_weight_credit: &mut XCMWeight,
+	) -> Result<(), ()> {
+		log::info!(
+			target: "xcm::barriers",
+			"AllowUnpaidExecutionFromCustom origin: {:?}, instructions: {:?}, max_weight: {:?}, weight_credit: {:?}",
+			origin, instructions, _max_weight, _weight_credit,
+		);
+		log::info!("origin {:?}", origin);
+		let incoming_parachain_id = V::get_incoming_parachain_id();
```

Comment:
I know this ticket is only for BRZ. But in the future, are we also thinking about supporting other tokens? From other parachains?

  For Amplitude, _`V`_ here is `MatcherConfigAmplitude`, which the func `get_incoming_parachain_id()` returns **_`MOONRIVER_PARA_ID`_**  all the time.

It seems very restrictive. 

@ebma @gianfra-t Please ignore this comment if that's what we're  aiming for.